### PR TITLE
[WIP] Fix MisteryMania

### PR DESCRIPTION
**Pull request by KiritoDv**
This PR tries to fix an old game called mistery mania, It implements this functions and classes:

- fs::write
- GenericChar and wchar::[strchr, strrchr, wcschr and wcsrchr]
- NSClass[initWithTitle and show]
- stdio::fgetc
- UIDevice::model
- NSData::[initWithBytes, initWithContentsOfFile, writeToFile and dataWithBytes]
- GLES::[glTexParameteriv, glTexSubImage2D, glBufferData, glColor4ub] and fixed null handling on glTexImage2D

### CODE REVIEW

**Comment by upintheairsheep** [comment]
Wow! Now talk about DEDICATION!!! +1
Also, this is a nice opportunity to try to test and add my code for even more extended UIDevice functionality. I'm honestly fine if you don't do this, and I won't annoy anyone for this off topic. https://github.com/hikari-no-yume/touchHLE/pull/78

**Comment by hikari-no-yume** [comment]
I've cherry-picked some of these now:

* d74edb530dbe1b739feb15a525c4141286820358 Implemented fs write function
* 1beaf35972cd3fb1e5a3733a9f71546a9f34adde Implemented strchr, strrchr, wcschr, wcsrchr
* d529dcf9e2ee6bdf1b1a473f97eb2889f322ce7f Implemented UIAlertView
* 34a39d69e2fb8659d48521b356053ba157c8adb1 Added UIDevice::model and hardcoded to iPhone
* bd678ea22426a54dcc9619eb2429513913443418 Implemented initWithBytes, initWithContentsOfFile, writeToFile and dataWithBytes

I made some small formatting clean-ups and also consolidated two of the commits, so the commits now on trunk aren't identical:

* 60cc53977275a4de0d972a18f29307b68cb59a3d Implement more NSData methods and added an fs write function
* 77e067bdcb678b77b23c62d0c682ca7cc7256cd5 Implemented stub for UIAlertView
* 0ead50643369d910dd7df3664ef5f3fefe350015 Added UIDevice::model and hardcoded to iPhone
* 17e9b66d3da0b4c468633f3e7d73c4a4ff256ad4 Implemented strchr, strrchr, wcschr, wcsrchr

I hope you don't mind. I figure it saves everyone time and effort if I just change some things myself, rather than requiring the submitter to. However, it will make merging a bit more complicated, sorry about that. You might want to do an interactive rebase instead (`git rebase -i`), then you can drop the conflicting commits, though that has its own risks.

As for the other patches: I haven't looked at all of them, but the ones I did look at seemed fine, but they'd need a little more tweaking and I can't be bothered right now. I hope I'll remember to look at these again later in the week.

**Comment by hikari-no-yume** [comment]
Since some of your work is now on trunk, I've credited you in the changelog using your GitHub username: 8174eebd3b88fb8fc060b2373dd90ffb96d31d33.

Please let me know if you'd prefer to be credited in a different way.

File: src/frameworks/opengles/gles_guest.rs [hunk 1]
```diff
@@ -523,9 +531,10 @@ fn glTexImage2D(
         };
         let pixel_count: GuestUSize = width.checked_mul(height).unwrap().try_into().unwrap();
         // This is approximate, it doesn't account for alignment.
-        let pixels = mem
+        let pixels = if pixels.is_null() { std::ptr::null() } else { mem
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles_guest.rs:534]: Have you run `cargo fmt` and `cargo clippy` before committing? I think it would've reformatted this.

File: src/frameworks/opengles/gles_guest.rs [hunk 2]
```diff
@@ -539,6 +548,51 @@ fn glTexImage2D(
...
+    xoffset: GLint,
+    yoffset: GLint,
+    width: GLsizei,
+    height: GLsizei,
+    format: GLenum,
+    type_: GLenum,
+    pixels: ConstVoidPtr,
+) {
+    with_ctx_and_mem(env, |gles, mem| unsafe {
+        let bytes_per_pixel: GuestUSize = match type_ {
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles_guest.rs:564]: This bit is shared with `glTexImage2D` and I think it always will be. Can you put it in its own function? Maybe it could be called `get_bytes_per_pixel`.

I would suggest doing something similar in `gles1_on_gl2.rs` as well.

File: src/frameworks/opengles/gles_guest.rs [hunk 3]
```diff
@@ -678,6 +732,20 @@ fn glDeleteRenderbuffersOES(env: &mut Environment, n: GLsizei, renderbuffers: Mu
     })
 }
 
+fn glBufferData(env: &mut Environment, target: GLenum, n: GLsizei, data: ConstVoidPtr, usage: GLenum) {
+    with_ctx_and_mem(env, |gles, mem| {
+        let data_ptr = mem.ptr_at(data.cast::<u8>(), 0).cast::<GLvoid>();
+        let n_usize: GLsizeiptr = n.try_into().unwrap();
+        unsafe { gles.BufferData(target, n_usize, data_ptr, usage) }
+    })
+}
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles_guest.rs:741]: These other functions look fine, but:

* Could you put them in separate commits from the texturing-related changes, since they seem unrelated?
* Could you try to put them near related functions in the file, and try to do this in all three GLES files? I notice that their position in the file doesn't match their position in the list of exported functions. Btw I think it'd make sense if `glBufferData` goes with the other functions related to buffers.

File: src/frameworks/opengles/gles_generic.rs [hunk 4]
```diff
@@ -9,7 +9,7 @@
 //! usage is to import `GLES` and `types` from this module, but get the
 //! constants from [crate::window::gles11].
 
-use crate::window::gles11::types::*;
+use crate::{window::gles11::types::*};
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles_generic.rs:12]: Can you revert this change?

**Review by hikari-no-yume — CHANGES_REQUESTED** [review]
This look good overall but I have some small suggestions.

Also, please try to keep the first line of the commit message to 72 characters so it doesn't get wrapped. You can add additional detail in the rest of the message (with an empty line inbetween it and the first line). If it's difficult to summarise what the commit does in 72 characters, that may be an indication it should be several commits instead, though of course this doesn't always apply.

File: src/frameworks/core_graphics/cg_image.rs [hunk 5]
```diff
@@ -61,6 +62,31 @@ impl HostObject for CGImageHostObject {}
...
+
+    let host_object = &env.objc.borrow::<CGDataProviderHostObject>(source);
+    let ptr: ConstPtr<u8> = Ptr::from_bits(host_object.data.to_bits());
+
+    let image = Image::from_bytes(&env.mem.bytes_at(ptr, host_object.size)).unwrap();
+    let host_obj = Box::new(CGImageHostObject {
+        image: image,
+    });
+
+    println!("CGImageCreateWithPNGDataProvider: {:?}", ptr);
```
Comment (hikari-no-yume) [inline src/frameworks/core_graphics/cg_image.rs:84]: This is a left-over debug print I guess?

File: src/libc/stdio.rs [hunk 6]
```diff
@@ -76,20 +76,15 @@ fn fgetc(
     let FILE { fd } = env.mem.read(file_ptr);
     let buffer = env.mem.alloc(1);
 
-    match posix_io::lseek(env, fd, 1, posix_io::SEEK_CUR) {
+    match posix_io::read(env, fd, buffer, 1) {
```
Comment (hikari-no-yume) [inline src/libc/stdio.rs:79]: What's going on here, is this an accident?

File: src/frameworks/opengles/gles_guest.rs (hunk 2 above)
Comment (hikari-no-yume) [reply src/frameworks/opengles/gles_guest.rs:564]: Actually, it could do the whole `ptr_at` thing too (without the null check).

**Comment by ghost** [comment]
@hikari-no-yume Hello? Any more updates on this PR?

**Comment by hikari-no-yume** [comment]
I've neglected it, to be honest, but I think @KiritoDv has been also been taking a break from this, so it's stalled. Why do you ask?

**Comment by hikari-no-yume** [comment]
Superseded by https://github.com/hikari-no-yume/touchHLE/pull/111.
